### PR TITLE
SYS-3338 convert nft manager storage to bounded

### PR DESCRIPTION
**Pull request by thadouk**
Updates the code, tests and Benchmarks to use a bound for external
references in storage.
Move constants into a separate module in avn-common crate.
Adds new tests to test the minting behaviour when the bound is not respected

### CODE REVIEW

**Review by alexpArtos — APPROVED** [review]

File: primitives/avn-common/src/event_types.rs [hunk 1]
```diff
@@ -345,7 +347,10 @@ impl NftMintData {
         // The actual unique ref is expected to be a UUID which is made up of 32bytes (WORD_LENGTH)
         // + 4 bytes for the dashes Example: b1dc0452-8b2f-78ec-7e80-167002d11678
         let ref_size = WORD_LENGTH + 4 * BYTE_LENGTH;
```
[the remark this answers is not in the document]
Comment (alexpArtos) [reply primitives/avn-common/src/event_types.rs:349]: I'm confused. The ExtRef limit specified in sp_avn_common seems to be 1024 bytes.
This ref_size is, if I'm not mistaken, 260 bytes.
And in the tests, we're using a 4096 for an arbitrary large size. Is this correct?

I'm not opposed to the first two. It means we have slack for other refs eventually. But I think the test should be aligned with the actual limit in the vector, and hopefully use the same constant plus 1.

Comment (thadouk) [reply primitives/avn-common/src/event_types.rs:349]: For the first two:
The external reference can be set from multiple interfaces, one of which is Ethereum which will utilise the equivalent of a UUID encoding in Ethereum (much less than 1024) see here:
- https://github.com/Aventus-Network-Services/avn-nft-tier1/blob/main/contracts/AvnNftListingsV1.sol#L423-L436
- https://github.com/Aventus-Network-Services/avn-nft-tier1/blob/main/contracts/AvnNftListingsV1.sol#L231

I can change the value for the tests into a +1 to the bound.

File: pallets/nft-manager/src/tests/single_mint_nft_tests.rs [hunk 2]
```diff
@@ -169,6 +169,28 @@ mod mint_single_nft {
...
+                let mut context: Context = Default::default();
+                let expected_info_id: NftInfoId = NftManager::next_info_id();
+
+                assert_eq!(false, <Nfts<TestRuntime>>::contains_key(&context.generate_nft_id()));
+                assert_eq!(false, <NftInfos<TestRuntime>>::contains_key(&expected_info_id));
+                assert_eq!(false, context.event_emitted_with_single_nft_minted());
+
+                let out_of_bounds_size = 4096;
+                context.unique_external_ref = (0..out_of_bounds_size).map(|_| 65).collect();
+
```
[the remark this answers is not in the document]
Comment (thadouk) [reply pallets/nft-manager/src/tests/single_mint_nft_tests.rs:185]: I can change this to use 1025 instead of 4096 to include the edge case. Any value outside the boundaries tests the boundary itself, let's not forget that.
Refactored it a bit to be easier to read, as I would like to avoid creating a function for 2 lines that initialise a variable to a value. It disrupts the flow when reading the code when you have to jump to a different area of the codebase to see that function implementation.

**Review by alexpArtos — APPROVED** [review]

**Review by nahuseyoum — APPROVED** [review]

**Review by ivan-cholakov — APPROVED** [review]

File: pallets/nft-manager/src/tests/single_mint_nft_tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexpArtos) [reply pallets/nft-manager/src/tests/single_mint_nft_tests.rs]: I agree with you. It does disrupt the flow. But that is where good naming comes in. You should not have to jump to the implementation when you first meet the function, you should trust it does what it says and carry on. That is where function extraction and good naming help in reading the code, you stay in the same place (as you would without extracting the function) and you stay at the same abstraction level (which is the improvement over _not extracting the function_).

Then, when you review the rest of the code, you'll eventually come across the new function. There, you just have to check two things: that the name properly corresponds to the implementation. Again, no jumping around: it's merely about checking the contract between functions is well-defined and well implemented.

As for the testing changes. I agree, any value outside the boundaries tests the changes, so with an infinity of values, which one do you choose? I try to choose _the first_ that triggers the behaviour I want to test. There is some trust, in here, in the monotonicity of the condition: if it triggers for value _n_, it will also trigger for value _n+1_. So, if the test is made for the first value that should trigger it, I believe that the behaviour will be correct for all the necessary cases. If I only do it for the 2000th case, then I have no tests for the space in between.

Ideally, what I like to do in my tests, is even more complete: test both sides of the boundary. If a condition should trigger on index _n_, I write two cases: 
* behaviour triggers on index _n_
* behaviour does not trigger on index _n-1_
Again, monotonicity is assumed.
